components/eamxx/src/physics/mam/impl/mam4_amicphys.cpp

```diff
@@ -1,1202 +1,1486 @@
+
+      qqcwsub[icnst][jclea] = 0;
+      EKAT_KERNEL_ASSERT_MSG(
+          fcldy != 0,
+          "Error! set_subarea_qmass_for_cldbrn_aerosols - fcldy is "
+          "zero\n");
+      qqcwsub[icnst][jcldy] = qqcwgcm[icnst] / fcldy;
+    }  // ispec - species loop
+  }    // imode - mode loop
+       //----------------------------------------------------------------
+}  // set_subarea_qmass_for_cldbrn_aerosols
+
+//--------------------------------------------------------------------------------
+//--------------------------------------------------------------------------------
+
+KOKKOS_INLINE_FUNCTION
+void get_partition_factors(const Real &qgcm_intrst,               // in
+                           const Real &qgcm_cldbrn,               // in
+                           const Real &fcldy, const Real &fclea,  // in
+                           Real &factor_clea, Real &factor_cldy)  // out
+{
+  //------------------------------------------------------------------------------------
+  // Purpose: Calculate the partitioning factors for distributing interstitial
+  // aerosol
+  //          mixing ratios to cloudy and clear subareas in a grid box.
+  //          The partitioning factors depend on the grid cell mean mixing
+  //          ratios of both interstitial and cloud-borne aerosols.
+  //------------------------------------------------------------------------------------
+
+  // qgcm_intrst  : grid cell mean interstitial aerosol mixing ratio
+  // qgcm_cldbrn  : grid cell mean cloud-borne aerosol mixing ratio
+  //
+  // fcldy        : cloudy fraction of the grid cell [unitless]
+  // fclea        : clear  fraction of the grid cell [unitless]
+  //
+  // factor_clea  : partitioning factor for clear  subarea
+  // factor_cldy  : partitioning factor for cloudy subarea
+
+  // Calculate subarea-mean mixing ratios
+
+  EKAT_KERNEL_ASSERT_MSG(fcldy != 0,
+                         "Error! get_partition_factors - fcldy is "
+                         "zero\n");
+  // cloud-borne,  cloudy subarea
+  const Real tmp_q_cldbrn_cldy = qgcm_cldbrn / fcldy;
+
+  // interstitial, cloudy subarea
+  const Real tmp_q_intrst_cldy =
+      haero::max(0, ((qgcm_intrst + qgcm_cldbrn) - tmp_q_cldbrn_cldy));
+
+  EKAT_KERNEL_ASSERT_MSG(fclea != 0,
+                         "Error! get_partition_factors - fclea is "
+                         "zero\n");
+  // interstitial, clear  subarea
+  const Real tmp_q_intrst_clea =
+      (qgcm_intrst - fcldy * tmp_q_intrst_cldy) / fclea;
+
+  // Calculate the corresponding paritioning factors for interstitial
+  // aerosols using the above-derived subarea-mean mixing ratios plus the
+  // constraint that the cloud fraction weighted average of subarea mean
+  // need to match grid box mean. Note that this subroutine is designed for
+  // partially cloudy grid cells, hence both fclea and fcldy are assumed to
+  // be nonzero.
+
+  constexpr Real eps = 1.e-35;  // BAD CONSTANT
```

Comment:
Currently, we are obsessed with keeping the answers as similar as we can. With all this code, we are still able to match very closely to double-precision answers from the Fortran codebase! We always add `BAD CONSTANTS` comment so that we can revisit these later. We may use numeric limits here from C++ in the future.